File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第8章方阵幂次.java
//方阵幂次 lanqiaoOJ 1551

import java.util.Scanner;
public class Main {
    static int n;
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        int q = sc.nextInt();   //q次幂
        int[][] A;
        A = new int[n][n];
        for(int i = 0;i < n; i++)
            for(int j = 0; j < n; j++)
                A[i][j] = sc.nextInt();
        int[][] res = pow_matrix(A,q);
        for(int i = 0; i < n; i++) {          //打印
            for (int j = 0; j < n; j++)
                System.out.print(res[i][j]+" ");
            System.out.println();
        }
    }

    static int[][] pow_matrix(int[][] a, int q) {  //快速幂
        int[][] res = new int[n][n];
        for(int i = 0; i < n; i ++)
            res[i][i] = 1;
        while(q > 0) {
            if(q % 2 == 1)       res = multi(res, a);
            a = multi(a,a);
            q /= 2;
        }
       return res;
    }
    static int[][] multi(int[][] a, int[][] b) {   //矩阵乘法
        int[][] temp = new int[n][n];
        for(int i = 0; i < n; i ++)
            for(int j = 0; j < n;j++ )
                for(int k = 0; k < n; k++)
                    temp[i][j] += a[i][k]*b[k][j];
        return temp;
    }
}

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第8章方阵幂次.py
#方阵幂次 lanqiaoOJ 1551

def multi(A, B):
    m1, n1 = len(A), len(A[0])
    m2, n2 = len(B), len(B[0])
    if n1 != m2: return None
    C = [[0] * n2 for i in range(m1)]
    for i in range(m1):
        for k in range(n1):
            for j in range(n2):
                C[i][j] += A[i][k] * B[k][j]
    return C

def power(A, n):
    N = len(A)
    res = [[0] * N for i in range(N)]
    for i in range(N): res[i][i] = 1
    while n:
        if n % 2:  res = multi(res, A)
        A = multi(A, A)
        n //= 2
    return res

s, q = map(int, input().split())
A = []
for i in range(s):  A.append(list(map(int, input().split())))
res = power(A, q)
for row in res:
    for c in row: print(c, end = ' ')
    print()
